airflow/models/dagrun.py

```diff
@@ -780,8 +780,7 @@ def _expand_mapped_task_if_needed(ti: TI) -> Iterable[TI] | None:
             except NotMapped:  # Not a mapped task, nothing needed.
                 return None
             if expanded_tis:
-                assert expanded_tis[0] is ti
-                return expanded_tis[1:]
+                return expanded_tis
```

Comment:
Since this function only returns _new_ ti objects, should we do something like this?

```python
if expanded_tis[0] is ti:
    return expanded_tis[1:]
return expanded_tis
```

Reply:
I explicitely return all here
Since there are 2 use cases for the `-1` instance:
- it might be deleted
- it might have a changed index

I now return all so I do not have to care about what happened on line 805.
If there was an expansion, I just add all so they will looped over.

If we only return new ones I would have to detect that on line 805 somehow.

I hope this makes the current way it works more clear.
If you want I could only return new ti objects, but I would need to figure out a way to see if the ti in the loop has been deleted yes or no

Reply:
If you prefer to keep the new behaviour (return all instead of only new), you need to modify the outer logic (naming, docstrings, and comments) to match.

Reply:
I changed the docstring of `_expand_mapped_task_if_needed`.

I think all the other comments and the naming is still fine.
If you think something needs to be changed just let me know :)